### PR TITLE
Fix LayerNorm instability issues.

### LINKED ISSUE #51102: LayerNorm does not behave as documented
## 🐛 Bug

torch.nn.LayerNorm does not behave as documented. It fails for me especially for highly biased/mean-shifted layers. I can't figure out where this instability yields from as it occurs even for weakly mean-shifted layers:

## To Reproduce
```
n = torch.nn.LayerNorm(4, eps=1e-5, elementwise_affine=False)
a = torch.randn(5, 4) + 2000  # create random tensor with N=5, L/C=4 and mean=2000, var=1
[[torch.mean(i), torch.var(i, unbiased=False)] for i in n(a)]  # should be near [0, 1]
```
gives:
```
Out[255]: 
[[tensor(0.0002), tensor(0.2006)],
 [tensor(0.0002), tensor(0.1634)],
 [tensor(0.0625), tensor(30822.9492)],
 [tensor(0.0001), tensor(0.0491)],
 [tensor(-0.2500), tensor(48425.8438)]]
```

## Expected behavior
```
def expected_LayerNorm(single_layer1d, eps):
    c = single_layer1d
    return (c - torch.mean(c)) / (torch.sqrt(torch.var(c, unbiased=False).add(eps)))

# comparison for one sample of a
print(n(a[2, :]))
print(expected_LayerNorm(a[2, :], 1e-5))
```
gives:
```
tensor([  22.0000,   72.7500, -285.0000,  190.5000])
tensor([ 0.1249,  0.4133, -1.6235,  1.0852])
```

## Environment
1.7.1+cpu (pip)
Win10

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @mruberry

### PR DESCRIPTION
**Pull request by ysiraichi**
Fixes #51102

Summary:
- `LayerNorm` instability issues were caused by instability when computing the variance
- Both kernels compute the squared sum which causes low precision for large values
- `torch.var` uses the Welford algorithm for computing the variance
- Changing the computation to the two-step algorithm solved the issue

### CODE REVIEW

**Comment by ysiraichi** [comment]
Would it be better to implement the Welford algorithm (better stability, but possibly slower), instead?

**Comment by ysiraichi** [comment]
If we wish to go with the welford algorithm, #54921 already implements it for CPU and GPU, so maybe we should wait for it?. That said, I also implemented it on CPU. The normalized shape for the results below are the biggest dimension.

| dtype | shape | [Two-Phase] Time (s) | [Welford] Time (s) |
| -- | -- | :--: | :--: |
| float32 | (500, 500) | 0.042 | 0.040 |
| float32 | (500, 1000) | 0.045 | 0.048 |
| float32 | (500, 2000) | 0.060 | 0.066 |
| float32 | (500, 4000) | 0.083 | 0.105 |
| float64 | (500, 500) |  0.042 | 0.048 |
| float64 | (500, 1000) | 0.057 | 0.067 |
| float64 | (500, 2000) | 0.092 | 0.103 |
| float64 | (500, 4000) | 0.133 | 0.174 |

Any thoughts?

File: aten/src/ATen/native/cuda/layer_norm_kernel.cu [hunk 1]
```diff
@@ -27,23 +27,37 @@ __global__ void RowwiseMomentsCUDAKernel(
...
     const int64_t index = i * N + j;
     sum1 += static_cast<T_ACC>(X[index]);
-    sum2 += static_cast<T_ACC>(X[index]) * static_cast<T_ACC>(X[index]);
   }
   sum1 = cuda_utils::BlockReduceSum<T_ACC>(sum1, m_shared);
+
+  if (threadIdx.x == 0) {
+      mean_ = sum1 * scale;
+  }
+  __syncthreads();
```
Comment (peterbell10) [inline aten/src/ATen/native/cuda/layer_norm_kernel.cu:48]: Is syncing actually faster than multiplying on each thread separately?

**Review by peterbell10 — COMMENTED** [review]
This is a much better way to calculate the variance but still could use a test that fails without it. Do you have any numbers comparing precision against welford ops on CPU and CUDA? Also, I think you should be benchmarking both CPU and CUDA. It's not clear what device your current benchmarks are showing, or even what code you used for welford ops.

File: aten/src/ATen/native/cuda/layer_norm_kernel.cu (hunk 1 above)
Comment (ysiraichi) [reply aten/src/ATen/native/cuda/layer_norm_kernel.cu:48]: The problem here is that after `BlockReduceSum`, only the first thread in the block has the fully reduced value of `sum1`. So, in my understanding, we would need to sync anyway in order to broadcast `sum1` to all threads in the block.

**Comment by ysiraichi** [comment]
In this PR, I still haven't implemented welford's algorithm.

Since #54921 (similar problem, but for `GroupNorm`) is already implementing it for both CPU and CUDA, I thought it would be a good idea to wait for that in case we want that algorithm. Let me try to summarize:

**Current implementation:**
```python
def var_1(x):
    mean = sum(x) / len(x)
    rstd = sum(i * i for i in x) # bad if 'i' is large
    return max(0, rstd / len(x) - mean * mean)
```
**PR implementation:**
```python
def var_2(x):
    mean = sum(x) / len(x)
    rstd = sum((i - mean) ** 2 for i in x) # bad if len(x) is large
    return max(0, rstd / len(x))
```

-----

- This PR does not implement Welford's algorithm, but solves the issue in #51102 (`i` was large)
- Apparently, Welford's algorithm is more robust (haven't ran any experiments, though), but is also less efficient
- [The benchmarks](https://github.com/pytorch/pytorch/pull/55044#issuecomment-812507979) illustrated the performance hit of a simple Welford implementation for CPU
- **Question:** should we wait for #54921 (implements the Welford algorithm for CPU and CUDA) be merged before moving on?

**Comment by ysiraichi** [comment]
Solved on #59987
